### PR TITLE
feat: Collapsible trees

### LINKED ISSUE #4: Collapse subtree
Me again, following #3, I would like to be able to collapse tree I already know (ex `@typescript-eslint`) to be able to look at the other dependency more easily.

### PR DESCRIPTION
**Pull request by hacknug**
This PR adds support for collapsing trees manually as well as replacement type (native, micro or both).

I wasn't sure how to expose those auto-filtering options so I decided to move the stats count you had in `<Hint>` out of the tooltip and merge them with the legend.

I also made some minor changes to the layout to improve the visibility of deeply nested branches on bigger screens as well as some minor to improve legibility and alignment of different elements.

There is an _issue_ that causes all of the sibling and children components to re-render when adding or removing a css class on an element. Not sure what would be the cleanest way to deal with this in React/Preact so I'd appreciate some guidance or help here.

Feel free to make or request changes as you see fit 👍

Closes #4.

### CODE REVIEW

File: src/components/Form.jsx [hunk 1]
```diff
@@ -43,7 +43,9 @@ export function PackageForm({ setQueryResult, fetchPkgTree }) {
 
     const onSubmit = async (e) => {
         e.preventDefault();
-        const pkgQuery = /** @type {string} */ (new FormData(formRef.current).get('pkgQuery')).trim();
+        const pkgQuery = /** @type {string} */ (
```
Comment (rschristian) [inline src/components/Form.jsx]: There's a few styling-only changes that should be reverted for starters. Diff should only contain necessary alterations please.

File: src/components/DataBox.jsx [hunk 2]
```diff
@@ -36,35 +36,73 @@ export function DataBox({ queryResult }) {
...
+    const [isNativeActive, setIsNativeActive] = useState(false);
+    const [isMicroActive, setIsMicroActive] = useState(false);
+
     return (
         <>
+            {!queryResult.error && (
+                <div class="z-10 sticky top-8 bg-card(& dark:dark) max-w-4xl mx-auto px-2 py-0 md:px-6 md:py-4 border(& resultBorder 1) rounded">
+                    <div class="grid gap-y-1 gap-x-4 grid-cols-1 sm:grid-cols-2 md:grid-cols-4 items-end [&>*]:p-1 [&>*]:text-left [&>*]:leading-none">
+                        <div
+                            class={[
```
Comment (rschristian) [inline src/components/DataBox.jsx]: I prefer interpolated class strings, i.e., ``class={`...`}``

File: src/components/DataBox.jsx [hunk 3]
```diff
@@ -74,18 +112,28 @@ export function DataBox({ queryResult }) {
...
-                        class="overflow-x-auto p-0.5"
+                        class={[
+                            'overflow-x-auto p-0.5',
+                            isNativeActive && isMicroActive
+                                ? 'collapse-tree-all'
+                                : isNativeActive
+                                ? 'collapse-tree-native'
+                                : isMicroActive
+                                ? 'collapse-tree-micro'
+                                : '',
```
Comment (rschristian) [inline src/components/DataBox.jsx]: 3-level ternary should probably be extracted to a separate var

File: src/components/PackageTree.jsx [hunk 4]
```diff
@@ -17,12 +21,33 @@ export function PackageTree({ pkg, depth = 0, isLast = false, prefix = '' }) {
         ? `underline(& offset-4) decoration(2 replacement-${pkg.replacement.type}(& dark:dark))`
         : '';
 
+    function handleTreeVisibility() {
```
Comment (rschristian) [inline src/components/PackageTree.jsx]: Should instead be defined inline, `onClick={() => setVisibility(...)}`

**Review by rschristian — COMMENTED** [review]
Did this from my phone, will take a better look when I'm at my desk later

**Comment by hacknug** [comment]
Thanks for the feedback! I've made changes to address your comments. Let me know if there's anything else I should do here.

I'll be out for the weekend so I'll probably be able to address them on Monday 👍

**Comment by rschristian** [comment]
Hey, sorry for the long time & no review, just been super busy this past week/week and a half. Really appreciate you donating time to this though.

I really, really like some of these changes but the general divergence in the UI is something I'm less of a fan of. Totally subjective of course, but I generally prefer to keep the UI pretty close to how it is at the moment rather than change it this much. As such, I'm going to try to take this time to extract out some of your features from here (will add you as a co-author to all work of course), probably doing the collapsing first and then the filtering.

That being said, I do want to note that I really love the collapse UI you built, it works so well and looks brilliant. I was struggling to come up with a good system for doing so, hence why that issue sat for so long, but what you've done is perfect. Seriously, well done & I really appreciate it.

So I don't think this PR will be mergable, feel free to close/keep open/whatever, but I absolutely will be going through the changes to try to land them in slightly different forms.

Sorry again for the delay, I do really appreciate the work.

**Comment by hacknug** [comment]
No worries! Glad I could help closing that issue 💖

**Pull request closed without merging**